packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol

```diff
@@ -0,0 +1,355 @@
+                    if gt(i, 0x00) {
+                        let x := mod(mload(0x00), gen_order)
+                        k := mulmod(k, x, gen_order)
+                        a := mulmod(a, x, gen_order)
+                        c := mulmod(challenge, x, gen_order)
+
+                        // calculate x_{j+1}
+                        mstore(0x00, keccak256(0x00, 0x20))
+                    }
+
+                    // Calculate the G1 element \gamma_i^{k}h^{a}\sigma_i^{-c} = B_i
+                    // Memory map:
+                    // 0x20: \gamma_iX
+                    // 0x40: \gamma_iY
+                    // 0x60: k_i
+                    // 0x80: hX
+                    // 0xa0: hY
+                    // 0xc0: a_i
+                    // 0xe0: \sigma_iX
+                    // 0x100: \sigma_iY
+                    // 0x120: -c
+                    calldatacopy(0xe0, add(noteIndex, 0x80), 0x40)
+                    calldatacopy(0x20, add(noteIndex, 0x40), 0x40)
+                    mstore(0x120, sub(gen_order, c)) 
+                    mstore(0x60, k)
+                    mstore(0xc0, a)
+
+                    // Call bn128 scalar multiplication precompiles
+                    // Represent point + multiplication scalar in 3 consecutive blocks of memory
+                    // Store \sigma_i^{-c} at 0x1a0:0x200
+                    // Store \gamma_i^{k} at 0x120:0x160
+                    // Store h^{a} at 0x160:0x1a0
+                    let result := staticcall(gas, 7, 0xe0, 0x60, 0x1a0, 0x40)
+                    result := and(result, staticcall(gas, 7, 0x20, 0x60, 0x120, 0x40))
+                    result := and(result, staticcall(gas, 7, 0x80, 0x60, 0x160, 0x40))
+
+                    // Call bn128 group addition precompiles
+                    // \gamma_i^{k} and h^{a} in memory block 0x120:0x1a0
+                    // Store result of addition at 0x160:0x1a0
+                    result := and(result, staticcall(gas, 6, 0x120, 0x80, 0x160, 0x40))
+
+                    // \gamma_i^{k}h^{a} and \sigma^{-c} in memory block 0x160:0x1e0
+                    // Store resulting point B at memory index b
+                    result := and(result, staticcall(gas, 6, 0x160, 0x80, b, 0x40))
+                    
+                    if eq(i, m) { // m = 0
+                        mstore(0x260, mload(0x20))
+                        mstore(0x280, mload(0x40))
+                        mstore(0x1e0, mload(0xe0))
+                        mstore(
+                            0x200,
+                            sub(0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47, mload(0x100))
+                            )
+                    }
+
+                    // If i > m + 1 (i.e. subsequent output notes)
+                    // then we add \sigma^{-c} and \sigma_{acc} and store result at \sigma_{acc} (0x1e0:0x200)
+                    // we then calculate \gamma^{cx} and add into \gamma_{acc}
+                    if gt(i, m) {
```

Comment:
m = 2 and n = 3, so this never gets called.

We should discuss in person but I think we should include both all 3 notes in the pairing check. Depending on whether this is a 'greater than' or 'less than' proof, the input and comparison notes get swapped around.

I think we should always include the comparison note in the range proof check, as the contract calling this range proof check may not have any agency over the structure of this note (it could be prover defined). So we should include all 3 notes in the pairing check.

Reply:
As discussed + reviewed, done